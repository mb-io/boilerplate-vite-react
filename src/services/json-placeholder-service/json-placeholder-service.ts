import axios, { AxiosRequestConfig, AxiosResponse } from 'axios'

import { axiosServiceErrorHandler } from '../axios-service-error-handler/axios-service-error-handler'

let requestCount = 0
let isThrottled = false

const resetThrottle = () => {
  requestCount = 0
  isThrottled = false
}

export interface JsonPlaceholderResponse {
  userId: number
  id: number
  title: string
  completed: boolean
}

export const jsonPlaceholderService = () => {
  const fetch = (options?: AxiosRequestConfig): Promise<JsonPlaceholderResponse> => {
    if (isThrottled) {
      throw new Error('Too many requests. Please wait a minute before trying again.')
    }

    if (requestCount >= 2) {
      isThrottled = true
      setTimeout(resetThrottle, 60000) // 1 minute
      throw new Error('Too many requests. Please wait a minute before trying again.')
    }

    requestCount += 1

    return axios
      .get(`/todos/1`, options)
      .then((response: AxiosResponse<JsonPlaceholderResponse>) => response.data)
      .catch((error) => {
        return axiosServiceErrorHandler(error, 'Failed to fetch placeholder data')
      })
  }

  return { fetch }
}
export type JsonPlaceholderService = AxiosResponse<JsonPlaceholderResponse>
