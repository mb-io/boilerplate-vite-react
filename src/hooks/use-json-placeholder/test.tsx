import { useJsonPlaceholder } from './use-json-placeholder'
import { jsonPlaceholderService } from '@/services/json-placeholder-service/json-placeholder-service'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { renderHook, waitFor } from '@testing-library/react'
import { Mock, vi } from 'vitest'

vi.mock('@/services/json-placeholder-service/json-placeholder-service')

const queryClient = new QueryClient()

describe('useJsonPlaceholder', () => {
  it('should fetch data from jsonPlaceholderService', async () => {
    const mockData = [{ id: 1, title: 'Test Post' }]
    const mockFetch = vi.fn().mockResolvedValue(mockData)
    ;(jsonPlaceholderService as unknown as Mock).mockImplementation(() => ({ fetch: mockFetch }))

    const wrapper = ({ children }: { children: React.ReactNode }) => (
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    )

    const { result } = renderHook(() => useJsonPlaceholder(), { wrapper })

    await waitFor(() => result.current.isSuccess)

    expect(result.current.data).toMatchSnapshot()
  })
})
