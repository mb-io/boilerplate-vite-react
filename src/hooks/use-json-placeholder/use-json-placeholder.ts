import { jsonPlaceholderService } from '@/services/json-placeholder-service/json-placeholder-service'
import { useQuery } from '@tanstack/react-query'

export const useJsonPlaceholder = () => {
  const service = jsonPlaceholderService()

  return useQuery({
    queryKey: ['json-placeholder'],
    queryFn: async () => await service.fetch(),
    enabled: true
  })
}
