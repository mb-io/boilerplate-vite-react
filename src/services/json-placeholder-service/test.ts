import { describe, it, expect, vi, beforeEach, afterEach, Mock } from 'vitest'
import axios from 'axios'
import { jsonPlaceholderService } from './json-placeholder-service'
import { axiosServiceErrorHandler } from '../axios-service-error-handler/axios-service-error-handler'

vi.mock('axios')
vi.mock('../axios-service-error-handler/axios-service-error-handler')

describe('jsonPlaceholderService', () => {
  let service: ReturnType<typeof jsonPlaceholderService>

  beforeEach(() => {
    service = jsonPlaceholderService()
    vi.useFakeTimers()
  })

  afterEach(() => {
    vi.clearAllTimers()
    vi.resetAllMocks()
  })

  it('should fetch data successfully', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    const result = await service.fetch()

    expect(result).toEqual(mockResponse.data)
    expect(axios.get).toHaveBeenCalledWith('/todos/1', undefined)
    expect(axios.get).toHaveBeenCalledTimes(1)
  })

  it('should handle errors using axiosServiceErrorHandler', async () => {
    const mockError = new Error('Network Error')
    ;(axios.get as Mock).mockRejectedValue(mockError)
    ;(axiosServiceErrorHandler as unknown as Mock).mockImplementation((error: Error, message: string) => {
      throw new Error(message)
    })

    await expect(service.fetch()).rejects.toThrow('Failed to fetch placeholder data')
    expect(axiosServiceErrorHandler).toHaveBeenCalledWith(mockError, 'Failed to fetch placeholder data')
  })

  it('should throttle requests after 2 attempts', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    try {
      await service.fetch()
      await service.fetch()
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    try {
      await expect(service.fetch()).rejects.toThrow('Too many requests. Please wait a minute before trying again.')
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    // Ensure the throttle is reset after 1 minute
    vi.runAllTimers() // Fast-forward time by 1 minute

    try {
      await service.fetch()
    } catch (error) {
      expect(error).toEqual(new Error('Failed to fetch placeholder data'))
    }
    expect(axios.get).toHaveBeenCalledTimes(1)
  })

  it('should reset throttle after 1 minute', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    try {
      await service.fetch()
      await service.fetch()
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    try {
      await expect(service.fetch()).rejects.toThrow('Too many requests. Please wait a minute before trying again.')
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    vi.runAllTimers() // Fast-forward time by 1 minute

    const result = await service.fetch()
    expect(result).toEqual(mockResponse.data)
    expect(axios.get).toHaveBeenCalledTimes(2)
  })

  it('should match the snapshot of fetched data', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    const result = await service.fetch()

    expect(result).toMatchSnapshot()
  })
})
