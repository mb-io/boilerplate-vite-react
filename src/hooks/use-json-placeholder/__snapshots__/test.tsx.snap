// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useJsonPlaceholder > should fetch data from jsonPlaceholderService 1`] = `undefined`;
